pkg/controller/perconaservermongodb/users.go

```diff
@@ -0,0 +1,261 @@
+		}
+	}
+
+	return restartSfs, nil
+}
+
+func (r *ReconcilePerconaServerMongoDB) updateUsersPass(cr *api.PerconaServerMongoDB, users []sysUser, adminUser, adminPass string, internalSysSecretObj *corev1.Secret) error {
+	for i, replset := range cr.Spec.Replsets {
+		if i > 0 {
+			log.Info("update users: multiple replica sets is not yet supported")
+			return nil
+		}
+
+		matchLabels := map[string]string{
+			"app.kubernetes.io/name":       "percona-server-mongodb",
+			"app.kubernetes.io/instance":   cr.Name,
+			"app.kubernetes.io/replset":    replset.Name,
+			"app.kubernetes.io/managed-by": "percona-server-mongodb-operator",
+			"app.kubernetes.io/part-of":    "percona-server-mongodb",
+		}
+
+		pods := &corev1.PodList{}
+		err := r.client.List(context.TODO(),
+			pods,
+			&client.ListOptions{
+				Namespace:     cr.Namespace,
+				LabelSelector: labels.SelectorFromSet(matchLabels),
+			},
+		)
+		if err != nil {
+			return errors.Wrapf(err, "get pods list for replset %s", replset.Name)
+		}
+		rsAddrs, err := psmdb.GetReplsetAddrs(r.client, cr, replset, pods.Items)
+		if err != nil {
+			return errors.Wrap(err, "get replset addr")
+		}
+		adminPass := string(internalSysSecretObj.Data[envMongoDBUserAdminPassword])
+		adminUser := string(internalSysSecretObj.Data[envMongoDBUserAdminUser])
+		client, err := mongo.Dial(rsAddrs, replset.Name, adminUser, adminPass, true)
+		if err != nil {
+			client, err = mongo.Dial(rsAddrs, replset.Name, adminUser, adminPass, false)
+			if err != nil {
+				return errors.Wrap(err, "dial:")
+			}
+		}
+		defer client.Disconnect(context.TODO())
+
+		for _, user := range users {
+			res := client.Database("admin").RunCommand(context.TODO(), bson.D{{Key: "updateUser", Value: user.Name}, {Key: "pwd", Value: user.Pass}})
+			if res.Err() != nil {
+				return errors.Wrap(res.Err(), "change pass")
+			}
+		}
+	}
+
+	return nil
+}
+
+func sysUsersSecretDataChanged(newHash string, usersSecret *corev1.Secret) (bool, error) {
+	secretData, err := json.Marshal(usersSecret.Data)
+	if err != nil {
+		return false, err
+	}
+	oldHash := sha256Hash(secretData)
+
+	return oldHash != newHash, nil
+}
+
+func sha256Hash(data []byte) string {
+	return fmt.Sprintf("%x", sha256.Sum256(data))
+}
+
+func (r *ReconcilePerconaServerMongoDB) restartStatefulset(cr *api.PerconaServerMongoDB, newSecretDataHash string) error {
+	for _, rs := range cr.Spec.Replsets {
+		sfs := appsv1.StatefulSet{}
+		err := r.client.Get(context.TODO(),
+			types.NamespacedName{
+				Namespace: cr.Namespace,
+				Name:      cr.Name + rs.Name,
+			},
+			&sfs,
+		)
+		if err != nil {
+			return errors.Wrapf(err, "failed to get stetefulset '%s'", rs.Name)
+		}
+
+		if sfs.Annotations == nil {
+			sfs.Annotations = make(map[string]string)
+		}
+		sfs.Spec.Template.Annotations["last-applied-secret"] = newSecretDataHash
+
+		err = r.client.Update(context.TODO(), &sfs)
+		if err != nil {
+			return errors.Wrapf(err, "update sfs '%s' last-applied annotation", rs.Name)
```

Comment:
I don't understand what does "last-applied annotation" here means. A bit confusing message as for me